Assets/_Scripts/Character/PlayerStats.cs

```diff
@@ -5,263 +5,407 @@
-	}
 
-	public void ResetSpeed() {
-		_speed = maxSpeed;
+	// Resetter ------------------//
+	public void ResetStats() {
+        foreach (PlayerStatWithModifiers stat in this.modifiableStats) {
+            stat.ClearModifiers();
+            stat.GetValue();
+        }
 	}
+
+    // Level up stat. Used for hero.
+    public bool LevelUpStat(Stat stat) {
+        PlayerStatWithModifiers theStat = this.modifiableStats.Find( (PlayerStatWithModifiers statWithMod) => statWithMod.stat == stat );
+        if (theStat == null) {
+            Debug.LogError("ERROR: Not assignable stat");
+            return false;
+        } else {
+           int costToLevelUp = StatLevelHelpers.GetCostToLevelUpStat(theStat.GetBaseValue());
+           if (costToLevelUp > this.statPoints) {
+               Debug.LogWarning("Not enough stat points to level up!");
+               return false;
+           }
+
+            this.statPoints -= costToLevelUp;
+            theStat.SetBaseValue(theStat.GetBaseValue() + 1);
+            return true;
+        }
+    }
+
+    // Applys level assuming that 
+    public void ApplyStatsBasedOnLevel(int level) {
+
+        if (level <= this.level) {
+            return;
+        }
+
+        statPoints += StatLevelHelpers.GetNumStatPointsForLevelUp(this.level, level);
+        this.level = level;
+
+        if (statPoints == 0) {
+          return;
+        }
+
+        List<ValueWeight<PlayerStatWithModifiers>> statsWithModifiersWeights = new List<ValueWeight<PlayerStatWithModifiers>>();
+        do {
+            statsWithModifiersWeights = new List<ValueWeight<PlayerStatWithModifiers>>();
+            foreach (PlayerStatWithModifiers statWithModifiers in this.modifiableStats) {
+                if (statPoints >=  StatLevelHelpers.GetCostToLevelUpStat(statWithModifiers.GetBaseValue())) {
+                    statsWithModifiersWeights.Add(new ValueWeight<PlayerStatWithModifiers>(statWithModifiers, statWithModifiers.originalBaseValue));
+                }
+            }
+
+            if (statsWithModifiersWeights.Count <= 0) {
+                break;
+            }
+
+            RandomPool<PlayerStatWithModifiers> statsPool = new RandomPool<PlayerStatWithModifiers>(statsWithModifiersWeights);
+            PlayerStatWithModifiers theStat = statsPool.PickOne();
+            int costToLevelUp = StatLevelHelpers.GetCostToLevelUpStat(theStat.GetBaseValue());
+            theStat.SetBaseValue(theStat.GetBaseValue() + 1);
+            statPoints -= costToLevelUp;
+        } while (statsWithModifiersWeights.Count > 0);
+
+        this.hp.SetValue(this.maxHp.GetValue());
+        this.mana.SetValue(this.maxMana.GetValue());
+
+    }
+
+    public void LevelUp() {
+        int additionalStatPoints = StatLevelHelpers.GetNumStatPointsForLevelUp(this.level, this.level+1);
+        this.statPoints += additionalStatPoints;
+    }
+
+    public List<PlayerStatWithModifiers> GetModifiableStats() {
+        return this.modifiableStats;
```

Comment:
Good practice to use `ReadOnlyCollection`

Reply:
Changed to dictionary instead